File: 42-trapping-rain-water/trapping-rain-water.java
class Solution {
    public int trap(int[] height) {
        int ans=0;
        int left_max=0;
        int right_max=0;
        int left=0;
        int right=height.length-1;
        while(left<right){
            if(height[left]<height[right]){
                left_max=Math.max(height[left],left_max);
                ans+=left_max-height[left];
                left++;
            }else{
                right_max=Math.max(height[right],right_max);
                ans+=right_max-height[right];
                right--;
            }
        }
        return ans;
    }
}

File: 42-trapping-rain-water/trapping-rain-water.py
class Solution:
    def trap(self, height: List[int]) -> int:
        ans = 0
        size = len(height)
        if size == 0:
            return ans
        left_cache, right_cache = [0] * size, [0] * size
        left_cache[0] = height[0]
        for i in range(1, size):
            left_cache[i] = max(left_cache[i - 1], height[i])

        right_cache[size - 1] = height[size - 1]
        for i in range(size - 2, -1, -1):
            right_cache[i] = max(right_cache[i + 1], height[i])

        for i in range(1, size - 1):
            ans += min(right_cache[i], left_cache[i]) - height[i]
        return ans


# [0,1,0,2,1,0,1,3,2,1,2,1]
